src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs

```diff
@@ -0,0 +1,147 @@
+{
+    public class AsyncSessionDocumentCounters : SessionCountersBase, ICountersSessionOperationsAsync
+    {
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, string documentId) : base(session, documentId)
+        {
+        }
+
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, object entity) : base(session, entity)
+        {
+        }
+
+        public async Task<Dictionary<string, long>> GetAllAsync()
+        {
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.GotAll)
+                {
+                    var result = new Dictionary<string, long>();
+                    foreach (var kvp in cache.Values)
+                    {
+                        if (kvp.Value.HasValue)
+                        {
+                            result.Add(kvp.Key, kvp.Value.Value);
+                        }
+                    }
+
+                    return result;
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache
+                {
+                    Values = new Dictionary<string, long?>()
+                };
+            }
+
+            cache.GotAll = true;
+            Session.CountersByDocId[DocId] = cache;
+            Session.IncrementRequestCount();
+
+            var vals = await Session.DocumentStore.Counters.ForDatabase(Session.DatabaseName).GetAllAsync(DocId).ConfigureAwait(false);
+            foreach (var kvp in vals)
+            {
+                cache.Values[kvp.Key] = kvp.Value;
+            }
+
+            return vals;
+        }
+
+        public async Task<long?> GetAsync(string counter)
+        {
+            long? value;
+
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.Values.TryGetValue(counter, out value) || cache.GotAll)
+                {
+                    return value;
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache
+                {
+                    Values = new Dictionary<string, long?>()
+                };
+                Session.CountersByDocId.Add(DocId, cache);
+            }
+
+            Session.IncrementRequestCount();
+            value = await Session.DocumentStore.Counters.ForDatabase(Session.DatabaseName).GetAsync(DocId, counter).ConfigureAwait(false);
+            cache.Values.Add(counter, value);
+
+            return value;
+        }
+
+        public async Task<Dictionary<string, long?>> GetAsync(IEnumerable<string> counters)
+        {
+            var countersToGetFromServer = new List<string>();
+            var result = new Dictionary<string, long?>();
+
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.GotAll)
```

Comment:
Merge the `GotAll` with the other call.
Then call to the server is didn't call `GotAll`

Reply:
fixed